compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1639,14 +1635,24 @@ private BType checkStructFieldAccess(BLangVariableReference varReferExpr, Name f
         BSymbol fieldSymbol = symResolver.resolveStructField(varReferExpr.pos, this.env,
                 fieldName, structType.tsymbol);
         if (fieldSymbol == symTable.notFoundSymbol) {
-            // check if it is an attached function pointer call
-            Name objFuncName = names.fromString(Symbols.getAttachedFuncSymbolName(structType.tsymbol.name.value,
-                    fieldName.value));
-            fieldSymbol = symResolver.resolveObjectField(varReferExpr.pos, env, objFuncName, structType.tsymbol);
+            if (structType.tag == TypeTags.OBJECT) {
+                // check if it is an attached function pointer call
+                Name objFuncName = names.fromString(Symbols.getAttachedFuncSymbolName(structType.tsymbol.name.value,
+                                                                                      fieldName.value));
+                fieldSymbol = symResolver.resolveObjectField(varReferExpr.pos, env, objFuncName, structType.tsymbol);
+
+                if (fieldSymbol == symTable.notFoundSymbol) {
+                    dlog.error(varReferExpr.pos, DiagnosticCode.UNDEFINED_STRUCT_FIELD, fieldName, structType.tsymbol);
+                    return symTable.errType;
+                }
+            } else {
+                // Assuming this method is only used for objects and records
+                if (((BRecordType) structType).sealed) {
+                    dlog.error(varReferExpr.pos, DiagnosticCode.UNDEFINED_STRUCT_FIELD, fieldName, structType.tsymbol);
+                    return symTable.errType;
+                }
 
-            if (fieldSymbol == symTable.notFoundSymbol) {
-                dlog.error(varReferExpr.pos, DiagnosticCode.UNDEFINED_STRUCT_FIELD, fieldName, structType.tsymbol);
-                return symTable.errType;
+                return ((BRecordType) structType).restFieldType;
```

Comment:
I guess you can remove LN 1660 as well, if so you can simply this logic, please check